lib/Parser/Parse.cpp

```diff
@@ -12749,15 +12819,20 @@ ParseNodePtr Parser::ParseDestructuredVarDecl(tokens declarationType, bool isDec
             }
         }
 
-        if (m_token.tk != tkID && m_token.tk != tkTHIS && m_token.tk != tkSUPER && m_token.tk != tkLCurly && m_token.tk != tkLBrack)
+        
+        if (m_token.tk != tkID && m_token.tk != tkTHIS && m_token.tk != tkSUPER)
         {
-            if (isDecl)
-            {
-                Error(ERRnoIdent);
-            }
-            else
+            bool nestedDestructuring = m_token.tk == tkLCurly || m_token.tk == tkLBrack;
+            if (isObjectPattern && nestedDestructuring || !isObjectPattern && !nestedDestructuring)
```

Comment:
You need some more parens here; clang doesn't like when you rely on order of operations for Boolean operators.